File: examples/PassThrough/client/cpp/Dynamic/PassThroughClient.cpp
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>
#include <stdlib.h>

#include "../gen-cpp/PassThroughService.h"

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

using namespace ::com::maxeler::PassThrough;

int check(std::vector<int32_t> dataOutDFE, std::vector<int32_t> dataOutCPU, int size) {
	int status = 0;

	for(int i=0; i < size; i++) {
		if(dataOutDFE[i] != dataOutCPU[i]) {
			fprintf(stderr, "Output data @ %d = %d (expected %d)\n",
				i, dataOutDFE[i], dataOutCPU[i]);
			status = 1;
		}
	}

	return status;
}

std::vector<int32_t> PassThroughCPU(int size, std::vector<int32_t> dataIn) {
	std::vector<int32_t> dataOut;
	dataOut.resize(size);

	for (int i=0 ; i<size ; i++) {
		dataOut[i] = dataIn[i];
	}

	return dataOut;
}

std::vector<int32_t> PassThroughDFE(int size, std::vector<int32_t> dataIn) { 
	// Make socket
	boost::shared_ptr<TTransport> socket(new TSocket("localhost", 9090));

	// Buffering is critical. Raw sockets are very slow
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));

	// Wrap in a protocol
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));

	// Create a client to use the protocol encoder
	PassThroughServiceClient client(protocol);

	const int sizeBytes = size * sizeof(int32_t);

	std::vector<int32_t> dataOut;
	dataOut.resize(size);

	try {
		// Connect!
		transport->open();

		// Initialize maxfile
		remote_ptr maxfile = client.PassThrough_init();

		// Load DFE
		remote_ptr engine = client.max_load(maxfile, "*");

		// Allocate and send input streams to server
		remote_ptr address_dataIn = client.malloc_int32_t(size);
		client.send_data_int32_t(address_dataIn, dataIn);

		// Allocate memory for output stream on server
		remote_ptr address_dataOut = client.malloc_int32_t(size);

		cout << "Running DFE.\n";

		remote_ptr actions = client.max_actions_init(maxfile, "default");
		client.max_set_param_uint64t(actions, "N", size);
		client.max_queue_input(actions, "x", address_dataIn, sizeBytes);
		client.max_queue_output(actions, "y", address_dataOut, sizeBytes);

		client.max_run(engine, actions);

		// Unload DFE
		client.max_unload (engine);

		// Get output stream from server
	        client.receive_data_int32_t(dataOut, address_dataOut, size);

		// Free allocated memory for streams on server
		client.free(address_dataIn);
        	client.free(address_dataOut);

		// Free allocated maxfile data
		client.PassThrough_free();

		// Close!
		transport->close();

	} catch (TException& tx) {
		cout << "ERROR: " << tx.what() << endl;
		exit(-1);
	}

	return dataOut;
}

int main() {
	// Input
	const int size = 1024;
	std::vector<int32_t> dataIn;

	for(int i = 0; i < size; i++) {
		dataIn.push_back(i + 1);
	}

	// CPU Output
	std::vector<int32_t> dataOutCPU = PassThroughCPU(size, dataIn);

	// DFE Output
        std::vector<int32_t> dataOutDFE = PassThroughDFE(size, dataIn);

	// Checking results
	if (check(dataOutDFE, dataOutCPU, size)) {
		cout << "Test failed." << endl;
		exit(-1);
	} else {
		cout << "Test passed!" << endl;
	}

	return 0;
}

File: examples/PassThrough/client/php/Dynamic/PassThroughClient.php
<?php 
set_include_path(get_include_path() . PATH_SEPARATOR . getenv('PHP_THRIFT_LIB'));

require_once 'Thrift/ClassLoader/ThriftClassLoader.php';

use Thrift\ClassLoader\ThriftClassLoader;

$GEN_DIR = '../gen-php';

$loader = new ThriftClassLoader();
$loader->registerNamespace('Thrift', getenv('PHP_THRIFT_LIB'));
$loader->registerDefinition('com\maxeler\PassThrough', $GEN_DIR);
$loader->register();

use Thrift\Protocol\TBinaryProtocol;
use Thrift\Transport\TSocket;
use Thrift\Transport\TBufferedTransport;
use Thrift\Exception\TException;

use com\maxeler\PassThrough\PassThroughServiceClient;

function check ($dataOutDFE, $dataOutCPU, $size) {
    $status = 0;

    for($i=0; $i < $size; $i++) {
        if($dataOutDFE[$i] != $dataOutCPU[$i]) {
            echo "Output data @ ".$i." = ".$dataOutDFE[$i]." (expected ".$dataOutCPU[$i].")\r\n";
            $status++;
        }
    }
        return $status;

}

function PassThroughCPU ($size, $dataIn) {
    for ($i = 0 ; $i < $size ; $i++) {
        $dataOut[$i] = $dataIn[$i];
    }

    return $dataOut;
}

function PassThroughDFE ($size, $dataIn) {
    $sizeBytes = $size * 4;
    try {
        $startTime = microtime(true);

        # Make socket
        $socket = new TSocket('localhost', 9090);

        # Buffering is critical. Raw sockets are very slow
        $transport = new TBufferedTransport($socket);

        # Wrap in a protocol
        $protocol = new TBinaryProtocol($transport);

        # Create a client to use the protocol encoder
        $client = new PassThroughServiceClient($protocol);
        
        echo "Creating a client:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Connect!
        $startTime = microtime(true);
        $transport->open();
        echo "Opening connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";
        
        # Initialize maxfile
        $startTime = microtime(true);
        $maxfile = $client->PassThrough_init();
        echo "Initializing maxfile:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Load DFE
        $startTime = microtime(true);
        $engine = $client->max_load($maxfile, "*");
        echo "Loading DFE:\t\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate and send input streams to server
        $startTime = microtime(true);
        $address_dataIn = $client->malloc_float($size);
        $client->send_data_float($address_dataIn, $dataIn);
        echo "Sending input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Allocate memory for output stream on server
        $startTime = microtime(true);
        $address_dataOut = $client->malloc_float($size);
        echo "Allocating memory for output stream on server:\t".(microtime(true) - $startTime)."s\r\n";

        # Action default
        $startTime = microtime(true);
        $actions = $client->max_actions_init($maxfile, "default");
        $client->max_set_param_uint64t($actions, "N", $size);
        $client->max_queue_input($actions, "x", $address_dataIn, $sizeBytes);
        $client->max_queue_output($actions, "y", $address_dataOut, $sizeBytes);
        $client->max_run($engine, $actions);
        echo "Pass through time:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";
 
        # Unload DFE
        $startTime = microtime(true);
        $client->max_unload($engine);
        echo "Unloading DFE:\t\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

        # Get output stream from server
        $startTime = microtime(true);
        $dataOut = $client->receive_data_float($address_dataOut, $size);
        echo "Getting output stream:\t(size = ".($size * 32)." bit)\t".(microtime(true) - $startTime)."s\r\n";

        # Free allocated memory for streams on server
        $startTime = microtime(true);
        $client->free($address_dataIn);
        $client->free($address_dataOut);
        echo "Freeing allocated memory for streams on server:\t".(microtime(true) - $startTime)."s\r\n";

        # Free allocated maxfile data
        $startTime = microtime(true);
        $client->PassThrough_free();
        echo "Freeing allocated maxfile data:\t\t\t".(microtime(true) - $startTime)."s\r\n";

        $startTime = microtime(true);
        $transport->close();
        echo "Closing connection:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

    } catch (TException $tx) {
        echo "ThriftException: ".$tx->getMessage()."\r\n";
        exit(-1);
    }

        return $dataOut;
}

$startTime = microtime(true);
$size = 1024;

for($i = 0; $i < $size; $i++) {
    $dataIn[$i] = $i + 1;
}
echo "Generating input data:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

# DFE Output
$startTime = microtime(true);
$dataOutDFE = PassThroughDFE($size, $dataIn);
echo "DFE pass through total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# CPU Output
$startTime = microtime(true);
$dataOutCPU = PassThroughCPU($size, $dataIn);
echo "CPU pass through total time:\t\t\t".(microtime(true) - $startTime)."s\r\n";

# Checking results
$startTime = microtime(true);
$status = check($dataOutDFE, $dataOutCPU, $size);
echo "Checking results:\t\t\t\t".(microtime(true) - $startTime)."s\r\n";

if ($status == 0) { 
    echo "Test successful!\r\n";
} else {
    echo "Test failed ", $status, " times!\r\n";
    exit (-1);
}
